extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPuller.java

```diff
@@ -89,18 +89,70 @@ public String getName()
       public InputStream openInputStream() throws IOException
       {
         try {
-          synchronized (inputStreamOpener) {
-            if (streamAcquired) {
-              return storageObject.getDataInputStream();
+          if (s3Object == null) {
+            synchronized (inputStreamOpener) {
+              if (s3Object == null) {
+                // lazily promote to full GET
+                s3Object = s3Client.getObject(objectSummary.getBucketName(), objectSummary.getKey());
+              }
             }
-            // lazily promote to full GET
-            storageObject = s3Client.getObject(s3Obj.getBucketName(), s3Obj.getKey());
-            final InputStream stream = storageObject.getDataInputStream();
-            streamAcquired = true;
-            return stream;
           }
+
+          return new InputStream()
+          {
+            final InputStream delegate = s3Object.getObjectContent();
+
+            @Override
+            public int read() throws IOException
+            {
+              return delegate.read();
+            }
+
+            @Override
+            public int read(byte[] b, int off, int len) throws IOException
+            {
+              return delegate.read(b, off, len);
+            }
+
+            @Override
+            public long skip(long n) throws IOException
+            {
+              return delegate.skip(n);
+            }
+
+            @Override
+            public int available() throws IOException
+            {
+              return delegate.available();
+            }
+
+            @Override
+            public void reset() throws IOException
+            {
+              delegate.reset();
+            }
+
+            @Override
+            public void close() throws IOException
+            {
+              delegate.close();
```

Comment:
`s3Object` should still be closed even if delegate close has an error

Reply:
Thanks, fixed.